### DISCUSSION TITLE
Should we support realms with whitespaces?

### DISCUSSION THREAD

**Discussion by DGuhr** [keycloak/keycloak#11978] [Ideas]
Hi everyone,

before going in the details, here's the tl;dr version:

### Situation
1. Today it is possible to `create` a realm where the name/id has a whitespace. This seemed to work in previous Versions of Keycloak (6.0 at least, see #11921 )
2. This has implications on the `baseUrl` and `redirectUri` of the client, and also on the `iss` field in the access token, as this does not get properly encoded/decoded.

### Conclusion 
Today it is possible to create a non-usable realm. This is not wanted. 

So two options I can see:
1) Forbid creating realms with whitespaces in the name. Possibly less invasive change.
2) Allow creating realms with whitespaces in the name, and fix the behaviour. This change is more invasive and touches some sensitive points (`iss`) of the OIDC Token. 

### Question(s) 
What option should we go with? Should we say "it is not supported"? Then we'd need to have client- and serverside checks when creating / importing a realm, with proper error messages to guide ppl. Also, this seems to be a regression (see #11921, answers from @djotanov ). OR: Should we add support for realms with whitespaces. This is a bigger change, and touches some sensible areas like the `iss` field parsing for the tokens in order to work correctly.

### Long version
While investigating a supposedly small problem with the new "Import at startup" feature in Quarkus based Keycloak in #11921, I discovered that this issue is also happening in Wildfly. 

Initial investigation in this Draft PR: #11964 showed that the problem was in the [validation](https://github.com/keycloak/keycloak/pull/11964/files/4bb7bdc5d72f12d43647b90944b9b87144676736#diff-9cd1d876021bd0401bc351dadb9e48c591bbd6f07f9935978f0839b90e194563) of the `ClientModel` that is imported, so first fix version was to use a simple `uri.replaceAll("\\s+","%20")` in the import.

The realm could get imported correctly. 

Not sure about the sideeffects here, I changed the draft PR #11964 to the current state, where it for now sanitizes the whitespaces to %20 before importing.

Then I tested creating a user and logging in and discovered, that while logging in to the account console worked, logging in to the admin console led to a 401 error. 

You can see it in action in the following video:
https://youtu.be/PrEzdw28j0Y

I also tested what happens with an OOTB Keycloak 18.0 installation, when you create such a realm with a whitespace in the UI. Result is you are not able to log in due to a wrong redirectUri set. You can see this in action in the following videos:
https://youtu.be/K9CbedIdgS0 / https://youtu.be/p5oOJOd_2ek

# Long story short
At the moment we have inconsistent and not functional behaviour when a realm with whitespace is created. Allowing this needs a bigger change, touching some more sensitive areas, so I wanted to reach out to ask you if we should go this path or if we should add the proper checks to not allow this at all instead.

@vmuzikar Feel free to add sth, and thanks for help investigating this :)

**Comment by andreaTP** [comment]
I think that:

> Forbid creating realms with whitespaces in the name

is completely fair, as it will step away from a number of real and possible issues.
The only attention would be to introduce the change in a `major` release and properly document it.

**Comment by djotanov** [comment]
If you decide to do that, please document migration path for clients that already have space in their realm. Looks like manual hacking in database would be required

**Reply by vmuzikar** [reply]
Hi @djotanov and thank you for your feedback! This would not most probably affect existing realms. However, we might hit validation issues when modifying such Realm.

@DGuhr Realm renaming was always a bit buggy. This will have to be properly tested.

**Reply by djotanov** [reply]
Realm updates is what I am talking about. We have automated process of realm upgrades, where we update realm JSON, then export old realm with users, replace old realm JSON with updated one, and reimport it back with overwrite. I'm sure many keycloak users are doing that, as there is no better way to do it at the moment (this might be simpler in the future with new store). 
This process will be broken (actually already is broken in newer keycloak versions) if you decide not to support whitespaces any more

**Reply by vmuzikar** [reply]
That's a good point. However, as far as I can tell, we never supported whitespaces properly. It is broken at multiple places (incl. token verification as Dominik stated). The suggestion is to improve the UX around it to make it clear whitespaces are not supported.

**Comment by abstractj** [comment]
Thanks for creating this discussion @DGuhr. We discussed this subject some time ago in a [couple of Jiras](https://issues.redhat.com/issues/?jql=project%20%3D%20KEYCLOAK%20AND%20text%20~%20whitespace). The amount of breaking changes and the effort required to support whitespace, from my perspective, does not worth it.

If we take URLs as an example, they are supported, but considered a bad practice at the same time. cc @keycloak/kc-developers

**Reply by pedroigor** [reply]
+1. We can improve it in terms of validation but not something we should rush (that is blocking/not working). Added some comments here too https://github.com/keycloak/keycloak/issues/11921#issuecomment-1127641525.

**Reply by mposolda** [reply]
+1 for not supporting whitespaces in realm names. At the same time, we can doublecheck if it makes sense to skip whitespaces also for some other things, which appears in the URLs and parameters (EG. client scope names, client ids etc).

The biggest issue with this change is the backwards compatibility though. I am not sure if we can handle DB updates for keycloak and also changes in realm JSON files. But I am pretty sure that we cannot handle everything and some manual steps will be needed by administrators (EG. Changing their `keycloak.json` adapter configurations to point to the changed realm name without whitespace etc).

Hence I guess we may need some deprecation period. Like for example:
- Step 1: Deprecate whitespaces in realm names. Which means skip allowing it by default, but introduce some config switch for backwards compatibilty
- Step 2 in some additional release: Remove whitespaces in realm names entirely

**Reply by vmuzikar** [reply]
+1 Makes total sense to me to make the validation disableable for some period of time.

**Reply by DGuhr** [reply]
ok, so it is pretty clear that we do not want to support this, makes total sense from my side, I just did not want to bias anyone by putting my opinion out first.  

@mposolda actually there is already a regression to older keycloak versions it seems, so a deprecation would not bring much imo, as it is unusable _as is_. See this video: https://youtu.be/3OhurnqPsLk

Even when you set the encoded whitespace manually, it does not work to access the account console ootb, or am I missing something? @djotanov mentions [here](https://github.com/keycloak/keycloak/issues/11921#issuecomment-1125968085), that it works for him. This seems a bit weird for me, as the video shows it does not, so -i guess I am missing sth here. If not, it is just broken imo, and even some checks are in place to avoid adding whitespaces when creating a client (see video), just not the check for when we create / import a realm. So _if I dont miss sth_ we don't need to deprecate it, but could just add the proper validation to the appropriate points. wdyt?

**Reply by DGuhr** [reply]
see also e.g. stians answer here: https://github.com/keycloak/keycloak/pull/5375#issuecomment-405343026 and pedros answer here and his referenced JIRA tickets: https://github.com/keycloak/keycloak/issues/11921#issuecomment-1127641525 - imo we should just remove the capability to create realms with whitespaces in it by adding proper validation, but I am fine when someone tells me we have to deprecate it first, I just need to know what I did wrong, because it seems completely broken to me, so no need to deprecate a broken feature ;)

**Reply by mposolda** [reply]
@DGuhr I've just tried to create keycloak realm called "foo bar" and it works fine. I am not able to login to account console by default, but after changing "Valid redirect URL" of the `account-console` client to `/realms/*` it works fine and I am able to login to the account console. Also people may not use account console at all and their clients can still work.

That is why I propose the "deprecation period" to be on the safe side regarding backwards compatibility.

**Reply by djotanov** [reply]
+1 for deprecation. We are currently able to upgrade our realm with whitespace to the latest keycloak version, and everything works fine (we don't use account pages).

**Reply by DGuhr** [reply]
@mposolda @djotanov thanks for clarifying, then i'd also say we need a deprecation period. Forgot about the wildcards, also it seems to work under some circumstances, so perhaps "account" console is not the best example. 

So @keycloak/kc-developers @keycloak/kc-maintainers question is how to proceed:
* assumption: It is clear we want to deprecate it.
* assumption: It is clear we want to use the existing validation in the server-spi-private package.
* fact: this change is not really quarkus/operator related, just evolved out of the referenced bug, and tbh I don't know yet all the places i'd have to touch (admin-ui old, admin-ui new, where to put the deprecated flag?, ...)
* fact: this change also needs to be communicated, targeted for a major version (imo: 19), added to migration guide, etc 
* Question: Anyone up for it? I'd do this and ask more concrete questions when I'm on it, but I'll be away the following week and as stated before may be not the best fit ;) (I'm still up for it if noone chimes in, don't get me wrong)

**Comment by kresnasatya** [comment]
I think whitespaces is not good and if developers create realm with 2 words with whitespace then replace whitespace with dash symbol or underscore symbol.

**Reply by vramik** [reply]
I believe that replacing whitespaces with underscores by default is not a way to go. I can imagine situations where caller could be surprised by this replacement. I think there should be rather validation done during the creation of the realm.

**Comment by mhajas** [comment]
One question related to this. Should we use realm names in URLs at all? Shouldn't we aim for referencing realms only by IDs and use names only in UI?

In my opinion, switching to IDs would also help with performance. For example, for some storage implementations, searching by ID can be better optimized than searching by some first level entity field (in this case realm name).

WDYT? cc. @keycloak/storage-x

**Reply by vmuzikar** [reply]
Good point. I think the Realm name and ID is the same thing in most cases. AFAIK we don't use UUID for Realm IDs(?).

**Reply by mhajas** [reply]
If I am not mistaken (@sguilhen could you please confirm), starting with https://github.com/keycloak/keycloak/pull/10901 we use UUIDs in the new storage JPA implementation.

**Reply by hmlnarik** [reply]
Realm names could be used in the URLs as long as they are unique - which should be the case - and as long as the lookup is done by realm name - which is the case (see https://github.com/keycloak/keycloak/blob/2e0d2ecbfb39e91a421f1142bf71366f85e230f5/services/src/main/java/org/keycloak/services/resources/RealmsResource.java#L203).
That said, no format of ID is guaranteed in the store - ID is store's responsibility, and store gives no guarantee apart from that it can be converted to and from `String`.

**Reply by vmuzikar** [reply]
>That said, no format of ID is guaranteed in the store - ID is store's responsibility, and store gives no guarantee apart from that it can be converted to and from String.

I think this is the answer to the question whether we should use IDs in the URLs. Since the IDs generation depends on the store implementation, we should use something more consistent – i.e. keep using the name.

**Comment by djotanov** [comment]
Removing support for spaces in realm names is all fine, as long as you provide instructions for users with existing realms with whitespace how to update their realm name. Updating realm name through the admin UI  currently does not work. The realm name gets updated, but the base URLs and redirect URLs remain the same in database. We would either need instructions for what manual database changes we would need to do, or fix the realm name update functionality.

**Reply by djotanov** [reply]
Actually update of realm name through the UI works as expected in latest version of keycloak, so that's fine

**Comment by stianst** [comment]
White-spaces (and other special characters) should not be permitted in Realm name by default. However, as there is a number of places where we should be more restrictive in what is permitted when adding more validation we need to consider backwards compatibility. For any new and more restrictive validation we should support disabling the validation. I don't want to add more bespoke validation though, and would like to address this as a part of better/unified validation in general.
